### DISCUSSION TITLE
How can I best learn WiX5 / MSI architecture?

### DISCUSSION THREAD

**Discussion by ChristianTellefsen** [wixtoolset/issues#8756] [Questions]
Hi WiX people!

I recently inherited a WiX3 solution which I plan to upgrade to WiX5 with the FireGiant VS2022 plugin. What's the best way to learn WiX5? Are there any online courses or guides that explain the architecture of WiX, how to develop solutions using WiX5, and so on? The examples I have found so far seem to be mostly "hello world" and how to click through a basic project with maybe a few extra bells and whistles. There was a nice web guide on FireGiant , but it did not go into great detail either.

I have never worked with WiX 3-5 and never made MSI installers, but know C#/.NET/XML and have 25+ years of professional experience coding all kinds of thingamabobs in various companies, so the guide can assume some general dev knowledge. :D 
 
Any pointers would be greatly appreciated!
 
 Cheers,
 Christian

**Comment by robmen** [comment]
What additional information are you looking for specifically?

**Comment by SJFriedl** [comment]
I'm in the same boat, it's been very painful: the windows installer system is exceptionally comprehensive, and good documentation is thin.

What's been super useful for me is ChatGPT: it knows quite about WIX and MSI, though it's not entirely up to speed on version 5. It's been particularly amazingly helpful on writing my custom action DLL (in C++), getting me skeleton code and helping me understand the custom action stuff in general.

Also: I imagine this would be nearly impossible to get right without some kind of sandbox / virtual machine: I use VMWare workstation, and I'm able to revert to the before-installer state before I try another turn. I think I read that Windows has some kind of sandbox mechanism that could be helpful here as well, but I don't know anything about it.

But there appears to be no shortcut to allocating a bunch of time and just slogging through it.

**Reply by bevanweiss** [reply]
Windows Sandbox is definitely the easier way to test the deployments.  It's much faster than reverting VMs to snapshots continually, and is also a bit 'safer' since by default the Windows Sandbox really has nothing installed inside of it, so you're unlikely to author an installer which has implied dependencies.

As for the actual Windows Installer stuff.. lots of it involves looking through the Microsoft Documentation for it, which is actually quite good (as far as the MS older documentation goes anyway).  It's a huge surface area to learn about in just the Windows Installer (MSI) territory, let alone things like additional WiX Custom Actions, or the whole world of bundles.
